### PR TITLE
feat: added support for custom nested menu entries in sidebar

### PR DESCRIPTION
**Pull request by semagarcia**
## Breaking change
No BC, this PR is retrocompatible

## Proposed change

Inside the sidebar menu, nowadays is not possible to group some entries for obtaining a better usability and better organization of all custom entries that we could add. In that sense, we could have a single sidebar menu entry called "Monitoring" and inside of it, we could have three links: Chronograf, Influx & Grafana. This avoids to have a huge sidebar menu:

image

These menus will be configured from the backend side, throught the configuration.yaml file, in order to change and customize it with ease (this will be contributed thru another PR against core).

The behavior of the new group/block for nested entries is shown in the following gif:

ha-sidebar-behaviour

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml
multilevel_sidebar_menu:
  direction: vertical
  entries:
    - menu_name: Configurator
      icon: mdi:pencil
      items:
        - title: File editor
          icon: mdi:file-edit-outline
          url_path: /url-for-file-editor
        - title: Visual Studio Code
          icon: mdi:microsoft-visual-studio-code
          url_path: /url-for-vscode
    - menu_name: Admin & Management
      icon: mdi:monitor-dashboard
      items:
        - title: Portainer
          icon: mdi:docker
          url_path: /url-for-portainer
    - menu_name: Monitoring
      icon: mdi:home-search-outline
      items:
        - title: Chronograf
          icon: mdi:database-cog
          url_path: /url-for-portainer
        - title: InfluxDB
          icon: mdi:chart-timeline-variant
          url_path: /url-for-influxdb-admin-app
        - title: Grafana
          icon: mdi:chart-areaspline
          url_path: /url-for-grafana-admin-dashboard-panel
```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

The doc will be added once the backend/core PR would be made

### CODE REVIEW

**Comment by karwosts** [comment]
Just to warn there's this older PR (#16017) working through which looks like a significant rewrite of the sidebar, there will be significant merge issues & re-writing for whoever doesn't get merged first.

**Comment by semagarcia** [comment]
> Just to warn there's this older PR (#16017) working through which looks like a significant rewrite of the sidebar, there will be significant merge issues & re-writing for whoever doesn't get merged first.

Thanks for your comment @karwosts and point any reader to that discussion from the PR! :)

**Comment by bramkragten** [comment]
I don't think the config for something like this should be in core, we have drag and drop support for changing the order of the sidebar, this should be part of that IMO.

**Comment by semagarcia** [comment]
Hi @bramkragten,

Let me clarify that this PR is not just for sorting options, because as you said, that it's already implemented. The proposed change is for enabling the capability of nested options inside the sidebar menu. Moreover, when any potential user use this capability, the component behaves as a block, because the order it's defined inside the config, but it's not the main idea.

The problem that I'd like to contribute is to add support for nested menus in the sidebar, because it either could grow easily or HA is accessed from tablets or mobile devices, having many entries would imply an unwanted scrollbar. So, we could maintain that shortcuts for quick and fast access, but in a more accessible and compact way.

Let me know if something it's not clear yet or if you need more details.

**Comment by matthiasdebaat** [comment]
Thanks for this PR @semagarcia! This is a nice feature for people with a lot of menu items. From a UX point of view I've some feedback:

- Groups can only be a label and not a menu item on itself. This means that clicking on the group label opens and closes it, never opening a page.
- In order for any user to be able to create groups, this must be done in the UI. We can use the smartphone pattern for this:
  - When you're in edit mode, you can use the drag and drop function. Dropping one menu item on top of another menu item creates a group with a default name and icon. You can change these inline in the edit mode.
  - Drag and drop is used to add and remove menu items from groups.
  - If there is no longer a menu item in the group, the group is deleted.
- Menu organisation is saved per device. We would need to find a way to store a default organisation per instance. 

If necessary, I can make a design to clarify things.

**Comment by semagarcia** [comment]
> Thanks for this PR @semagarcia! This is a nice feature for people with a lot of menu items. From a UX point of view I've some feedback:
> 
> * Groups can only be a label and not a menu item on itself. This means that clicking on the group label opens and closes it, never opening a page.
> * In order for any user to be able to create groups, this must be done in the UI. We can use the smartphone pattern for this:
>   
>   * When you're in edit mode, you can use the drag and drop function. Dropping one menu item on top of another menu item creates a group with a default name and icon. You can change these inline in the edit mode.
>   * Drag and drop is used to add and remove menu items from groups.
>   * If there is no longer a menu item in the group, the group is deleted.
> * Menu organisation is saved per device. We would need to find a way to store a default organisation per instance.
> 
> If necessary, I can make a design to clarify things.

I understand your feedback @matthiasdebaat, but let me ask you some questions:

1. With your proposal, you're working entirely from the frontend side; I realize the purpose of that in terms of accessibility, ease of use and simplicity, but, how do you configure the URL of each entry (or other properties like "just for admin" or "enabled users") if you are adding it/them "manually"? Because we could have custom entries (like ones coming from _panel iframe_ component) that needs the target href. Even we could evolve this component allowing to open the nested menu horizontally in case of tablet/desktop devices in order to reduce the presence of the scrollbar.
2. Why do yo think that would be necessary to implement these behavior without any config file (or addressed by that config file)? Could we propose an initial release based on the previously mentioned config file and behavior, and in a next release (typically, v2 for instance) implement that improvement? We could take into consideration the feedback of the initial users too.
3. I think that the configuration could be saved inside the own configuration.yaml one, in order to be possible to backup it, restore it and saved the preferences of the user for that instance easily, because that feature -if approved- will be a built-in capability.

Let me know if some doubt isn't clear, I'll be proud to answer you (or anyone would be interested in the discussion).

**Comment by bramkragten** [comment]
The panels you are grouping in the nested menus already exist right? We are not creating new panels in this PR?

A first version should have limited functionality so creating a group and adding/removing an item from/to that group. Every other functionality should not be part of the first version.

The problem with using `YAML` to configure this, is that we can never write `YAML`, only read it, so we would never be able to add a UI editor for this feature if the config is in `YAML`, meaning it would be a breaking change/difficult migration for the user is we would add an UI editor, because we then have to store the config in `JSON`, and way more complicated code to handle both cases.

**Comment by semagarcia** [comment]
- "The panels you are grouping in the nested menus already exist right? We are not creating new panels in this PR?" => Yes, we are creating then, because the initial goal of this PR was just grouping (nesting) shortcuts/links to existing pages (such as grafana dashboards, portainer...), and that's the reason why we need to specify which is the target URL of each subitem. At the beginning, I was inspired in the panel iframe component to following a similar approach.
- "A first version should have limited functionality so creating a group and adding/removing an item from/to that group. Every other functionality should not be part of the first version" => So, if a user wants to create a new menu entry with two links pointing to portainer & grafana, how it should be done from your point of view in that basic functionality? For instance, that shortcuts are not added directly from HA (as long as I know), but also you need other plugin/components to do that as the previous iframe component mentioned.
- "
The problem with using YAML to configure this, is that we can never write YAML, only read it, so we would never be able to add a UI editor for this feature if the config is in YAML, meaning it would be a breaking change/difficult migration for the user if we would add an UI editor, because we then have to store the config in JSON, and way more complicated code to handle both cases" => I'm not know HA internals an its core deeply (althought I'd like), but I guess that there is a mechanisim to save some data, like user preferences, dashboard -lovelace- configurations ... so, which would be the recommendation to save that data (I mean, the new menu arrangement)? Is there any guideline that I could read/follow to try to refactor the PR to accomplish that? And, if we dismiss the use of a YAML/JSON, the steps should be: create a new sidebar item (for instance, through iframe panel component) and once it has been added, drag-and-drop to add or remove it, right? And in that case, how could I rename that new created group with a default name? And the last one... how should be the behavior if a user already uses the custom panel, because that component reads its config from YAML (converted to JSON) in the context of the rest of sidebar entries.

**Comment by semagarcia** [comment]
> * "**The panels you are grouping in the nested menus already exist right? We are not creating new panels in this PR?**" => Yes, we are creating then, because the initial goal of this PR was just grouping (nesting) shortcuts/links to existing pages (such as grafana dashboards, portainer...), and that's the reason why we need to specify which is the target URL of each subitem. At the beginning, I was inspired in the panel iframe component to following a similar approach.
> * "**A first version should have limited functionality so creating a group and adding/removing an item from/to that group. Every other functionality should not be part of the first version**" => So, if a user wants to create a new menu entry with two links pointing to portainer & grafana, how it should be done from your point of view in that basic functionality? For instance, that shortcuts are not added directly from HA (as long as I know), but also you need other plugin/components to do that as the previous iframe component mentioned.
> * "
>   **The problem with using YAML to configure this, is that we can never write YAML, only read it, so we would never be able to add a UI editor for this feature if the config is in YAML, meaning it would be a breaking change/difficult migration for the user if we would add an UI editor, because we then have to store the config in JSON, and way more complicated code to handle both cases**" => I'm not know HA internals an its core deeply (althought I'd like), but I guess that there is a mechanisim to save some data, like user preferences, dashboard -lovelace- configurations ... so, which would be the recommendation to save that data (I mean, the new menu arrangement)? Is there any guideline that I could read/follow to try to refactor the PR to accomplish that? And, if we dismiss the use of a YAML/JSON, the steps should be: create a new sidebar item (for instance, through iframe panel component) and once it has been added, drag-and-drop to add or remove it, right? And in that case, how could I rename that new created group with a default name? And the last one... how should be the behavior if a user already uses the custom panel, because that component reads its config from YAML (converted to JSON) in the context of the rest of sidebar entries.

Hi @bramkragten , any thoughts about this? Thx!

**Comment by piitaya** [comment]
For now, the sidebar configuration is stored in browser local-storage. This means it's not back-end persisted. 
We should add a way to store the data at the server level so every user can have the same sidebar configuration.

**Comment by jlpouffier** [comment]
Hello 👋 

We discussed this with the team yesterday and I'll try to give here the outcome of this discussion.

_TL;DR: This is not the right time to push this PR because we do not know where we want to go with our sidebar. We do not know where we want to go with the sidebar because we do not know what all our users want to accomplish_

We all agree that our sidebar needs some love.

We would like to allow our users to be able to put what they want on their sidebar.
However, what we want to avoid is pushing features too soon while running the risk of being stuck later on if we decide to go on a different path.
_(As you know, what comes into Home Assitant usually stays forever, with rare exceptions)_

While nesting entries is definitely something that resonates with you, we still don't know the size of the group that would like to have nested entries.
More importantly, we do not know what dependencies it may bring later on if we decide to change course.

For transparency, the main questions we raised internally were:
- Shall we store the sidebar config at the server level? At the user level? Or at the device level locally as of today?
- How do users use the sidebar today? What are their struggles? What are they trying to achieve?
- Shall we allow users to put any item in the sidebar (Some hidden entry of the settings menu for example Zone or Analytics, some external URL like your favorite news page?)

What we agreed on was to start, later on, a small research effort in order to understand how users use the sidebar, and what they are trying to accomplish.
We would like to keep you in the loop when we do this because you seem to have a lot of ideas, and we definitely want to tap into them 😊

Feel free to comment, we remain available!

**Comment by semagarcia** [comment]
First of all, thanks for your detailed answer @jlpouffier explaining and detailing the reasons to close it; I understand them and I agree with you.

For me it will be a pleasure to have the possibility to collaborate with such a great open-source project contributing with that feature in the future. As any other IT project, developers need to listen their users and what they want. Totally reasonable.

Meanwhile, I was thinking to try to release the initial behavior mentioned (the proposed one through this PR) like a plugin if someone could find it interesting or some reader reaches this conversation before the built-in feature is implemented.

Don't hesitate to make me any comment!
Thanks!

**Comment by stevland** [comment]
I love this idea. I found @semagarcia's merge request while looking for this exact functionality. It's essential.

As @bramkragten points out, it would need to be integrated with HA's drag and drop sidebar ordering capability.

Forgive me if this comment seems really trivial. I only want to point out that WordPress has had both of these capabilities for a really long time. 

animated-03-low

It could serve as an excellent example to emulate.

Good luck, @semagarcia. I hope your contribution is reconsidered soon!

**Comment by m32w** [comment]
I would love to have this feature in 2024.2

**Comment by potelux** [comment]
It is a great feature, but as mentioned before, it will not be added right now as the future of the sidebar is unknown.  

Keep an eye on the various PRs regarding"labels" as they may prove helpful for this in the future.
